### DISCUSSION TITLE
{{selection}}: Automatic selecting if nothing is selected

### DISCUSSION THREAD

**Discussion by Taitava** [Taitava/obsidian-shellcommands#54] [Ideas]
This is just a very quick idea that I haven't thought through yet. If there is a caret in the text (in editor mode), but nothing is selected, a command could define what to do:
- `{{selection:select-word}}`: Highlight the current word at the caret (only if nothing is selected).
- ~~`{{selection:select-row}}`~~`{{selection:select-line}}`: Highlight the current line at the caret, adhering to soft wraps. (only if nothing is selected).
- `{{selection:select-I-DONT-KNOW-YET}}`: Highlight the current line at the caret, adhering to hard linebreaks. (only if nothing is selected).
- `{{selection:select-paragraph}}`: Highlight the current paragraph at the caret (only if nothing is selected).
- `{{selection:select-all}}`: Highlight the whole content in the note (only if nothing is selected).
- `{{selection:select-all-but-yaml}}`
- `{{selection:select-yaml}}`
- `{{selection:select-link}}`: Highlight the current link at the caret (only if nothing is selected). (Thanks, @FelipeRearden!)
- Other ideas?

This could be good for some e.g. *Use this as input and replace it with output* tasks, and you don't have to always highlight text beforehand, just point the caret at a correct word. Would there be any use for something like this?

---

Another feature, added 2022-03-02:

## Ability to cancel execution if nothing is selected

This is completely different than the "select" preaction described above. It could work with some events (prevent automatic execution if no text is selected), and also prevent accidental hotkey/command palette originated execution in case a user accidentally tries to execute a shell command without having anything selected, and if the shell command is not designed to work properly with an empty selection.

There would be an option to either cancel silently or with an error message.

Another solution to this might be to change the `{{selection}}` variable to raise an error if nothing is selected. Then with the possibility to define default values for variables (will be implemented together with _custom variables_ #146 ), the variable can be defined to allow execution even without a selection, if needed.

Edit 2022-12-10: This has now a separate discussion: #299

**Comment by FelipeRearden** [comment]
I loved everything !!!! That's a great variable. 

Text transportation is something that is important too me. 

> {{selection:select-word}}: Highlight the current word at the caret (only if nothing is selected).

Select the word -> copy to clipboard -> open Search Plug-in via URI -> Shell Command to focus on text field -> paste into the search field :)

 > {{selection:select-row}}: Highlight the current row at the caret (only if nothing is selected).

Obsidian users use the term line more often than Row :)

> {{selection:select-paragraph}}: Highlight the current paragraph at the caret (only if nothing is selected).

Would mind sharing an example of a paragraph ? I wanna see if is the same concept that I'm thinking (lines/row that aren't separated by an empty line) 

> {{selection:select-all}}: Highlight the whole content in the note (only if nothing is selected).

This variable was in my mind for some time. 
![image](https://user-images.githubusercontent.com/85180696/135329038-3de3f7b9-f767-4cb6-95d3-34b22f34ac25.jpeg) 

> Other ideas? 

I will bring more :)


Have a great day!!!!!

**Comment by Taitava** [comment]
> > {{selection:select-row}}: Highlight the current row at the caret (only if nothing is selected).
> Obsidian users use the term line more often than Row :)

Oh, true, _line_ is better!

> > {{selection:select-paragraph}}: Highlight the current paragraph at the caret (only if nothing is selected).
> Would mind sharing an example of a paragraph ? I wanna see if is the same concept that I'm thinking (lines/row that aren't separated by an empty line)

I haven't actualle checked how easy it will be in Obsidian API to access these different elements such as line and paragraph, but at least now my idea is that *paragraphs* would mean text blocks separated by empty space between them, so a single linebreak would not be enough to separate paragraphs. Technically a *paragraph* would refer to HTML's `<p>` element.

Maybe there would need to be another option between *line* and *paragraph*: something that selects from previous linebreak chracter to next linebreak character.

> > {{selection:select-all}}: Highlight the whole content in the note (only if nothing is selected).
> This variable was in my mind for some time.

`{{NoteContent}}` is closely related, but also a bit different. `{{selection:select-all}}` has the option that it can give something else than the whole content, in case a user has something selected already. Here is a new discussion for `{{NoteContent}}`: https://github.com/Taitava/obsidian-shellcommands/discussions/57

`{{selection:select-all}}` clearly needs an additional version that would not select YAML headers. And yet another, that only selects the YAML part.

**Reply by FelipeRearden** [reply]
> I haven't actualle checked how easy it will be in Obsidian API to access these different elements such as line and paragraph,

This is gonna be a great variable, you could bring something really new to our community. Let me know in the future because I think we could have many types of paragraphs to play with :)

**Comment by FelipeRearden** [comment]
> Other ideas?

`{{selection:select-link}}`: Highlight the current link at the caret (only if nothing is selected).
- internal link `[[ ]]`
- embed link `![[ ]]`
- markdown link `[]()`

**Reply by Taitava** [reply]
Good idea, thanks!

**Comment by Taitava** [comment]
Now that I have been thinking about this for a while, I'm not so certain if this would be the best way to implement this. The ability to select text before executing a command is a good idea per se, and I would like to find some form to implement it.

But I'm not confident about implementing this in the form of a variable. So far, `{{variable}}`s have been made to **read** values without side effects. It could be used as a design principle, too.

Maybe a better idea would be to implement some kind of _preactions_: select something if nothing is selected, activate last active file if no file is active, activate a certain file if no file is active, always activate a certain file etc. These would be completely separated from variables. And I should start a new conversation for this, and then edit this current conversation's title so that it talks about a new *Preaction* for selecting text, rather than about editing `{{selection}}` variable.

**Reply by FelipeRearden** [reply]
@Taitava , if I understood correctly you are talking about a techincal change rather than a concept/principles change (sorry if I didn't understood) :)

Two questions:

**_Preaction_ is a set of "actions" that will be chosen by the user when creating a SC?**

- {{select-word}}
- {{select-line}}.
- {{select-I-DONT-KNOW-YET}}
- {{select-paragraph}}
- {{select-all}}
- {{select-all-but-yaml}}
- {{select-yaml}}
- {{select-link}}

**The _Preaction_ will work together with the `{{selection}}` in the Shell Command ?** 

Example: I want select the active and send to _MyNote.md_

`echo {{selection}} >> MyNote.md [preaction={{select-word}}`

**Reply by Taitava** [reply]
Yes, you understood correctly. *Preaction* can be selected by the user. There will be a preaction named *select* (or something) and it will have a dropdown menu to further indicate what will be selected (any of the options listed above). The *select* preaction is just one example of preactions. There will be other preactions too, and a user is able to select multiple preactions to perform before shell command execution.

> echo {{selection}} >> MyNote.md [preaction={{select-word}}

Preactions will be defined in a clickable user interface, so there's no need to type `[preaction={{select-word}}` to the actual shell command, but you understood the general point very well! 🙂

**Reply by FelipeRearden** [reply]
> Yes, you understood correctly.

I'm getting smarter talking to you :)

> Preaction can be selected by the user. There will be a preaction named select (or something) and it will have a dropdown menu to further indicate what will be selected (any of the options listed above).

> Preactions will be defined in a clickable user interface, so there's no need to type [preaction={{select-word}} to the actual shell command

Great ideas for the UI !!!!

> The select preaction is just one example of preactions. There will be other preactions too, and a user is able to select multiple preactions to perform before shell command execution.

That's what I thought !!!! This gonna be AMAZING 🤩

**Reply by Skefflock** [reply]
Hi, I may suggest a couple more options for this feature.

It seems reasonable to me to have an option to replace text under some certain header. 
Also to add output to the end of a text under some header. This will allow to add information to a specific place in note or move some piece of information to a new destination in current or another note. 

And also to select a paragraph not at the caret position but starting from a position of some marker like tag or a sequence of symbols.

**Reply by Taitava** [reply]
Hi! Thank you for your suggestion. If I understood correctly, this is what you are suggesting?
- _"replace text under some certain header"_ = select all text in a certain chapter? I.e. Chapter meaning here all text under a heading, until next heading or end of file. Then, when output is received, the selected text will be replaced with the new output.
- _"to add output to the end of a text under some header"_ = no selecting, but just place the caret after all text under a certain heading. Then, when output is received, new output is inserted at the caret position, i.e. at the end of the chapter.

> And also to select a paragraph not at the caret position but starting from a position of some marker like tag or a sequence of symbols.

So, if e.g. a paragraph contains my name (`Jarkko`), or whatever sequence of characters, that particular paragraph would be selected?

I think these ideas sound reasonable, I'll just need to think about them so that I'll get a clear picture on what is the most simplest way to design them. (simple = simple to implement, but also simple to use).

**Comment by Skefflock** [comment]
Yep, this is all correct. Thanks.

**Reply by Taitava** [reply]
Nice. I'll continue thinking about this and what support Obisidan API has for these kinds of features. 🙂
